### DISCUSSION TITLE
Work In Progress

### DISCUSSION THREAD

**Discussion by robmoffat** [risk-first/website#78] [Team Posts]
So a couple of months back I started working on a few new articles about estimating.  I wanted to try and write up something on the different approaches to estimating software development, and the limits to our ability to estimate. 

At the same time, I had some ideas about complexity, and writing up the main tools we use for managing complexity in a code-base.  

But the whole thing has turned into a bit of a rabbit-hole, and both of these concepts ended up being linked together.  

I still have a long way to go on this.  Nevertheless, I thought I'd share the first five articles on here so that I can get some feedback, and see what you guys think.  

Many apologies - this is not as clear or polished as I'd usually like, but I think over the next month or so it should shape up a bit.

https://riskfirst.org/Estimating

cheers,
Rob

**Comment by bretweinraub** [comment]
Have you ever looked at McConnell's "Software Estimation"?

**Comment by robmoffat** [comment]
Hey Bret, 

Short answer, no... and I'm leery of shelling out again for one of his books on Amazon as it's a) 13 years old and b) I didn't get the value out of the last one I bought of his.  

Is there something specific that you want me to know?  From this post, https://stevemcconnell.com/blog/17-theses-software-estimation/, this is all very predictable software estimation fare:  measuring past projects, measuring velocity etc.

Having said that, he seems to have been having an interesting discussion with Ron Jeffries about [estimation lately](https://stevemcconnell.com/blog/noestimates-response-ron-jeffries/). Although from the look of it, they're not making much headway!  In this, they're basically discussing whether estimates are or aren't useful, and in what context.  

I'm really hoping I can dig a bit deeper than this.  That's why I'm trying to look at the axes along which projects lie and figure out the domain in which classical extrapolation works, and how it fails.   

Personally, I now have a feel for certain _types_ of task, where I can tell how long they're going to take.  And I also have some kind of alarm-bell for the types of task that are going to go off the rails.  Based mainly on bitter experience.

I'm trying to explore this... but if McConnell already has something around this then let me know.

**Comment by bretweinraub** [comment]
I've got it in hard copy ..... somewhere?   I might have it on a kindle.

**Comment by bretweinraub** [comment]
What about the risk that whoever asks you to pull an estimate considers it
a quote :)

**Comment by robmoffat** [comment]
You mean where you give an estimate and suddenly all the risk of the project is on you?  Hold up, I think I started a diagram for that!

![Accepting An Estimate](https://riskfirst.org/images/generated/estimating/accept_estimate.png)

Seriously though, that's where things start getting interesting.  I haven't written about that yet, I'm still in the foothills.  But that's seriously fun territory.   Here's a big question:  does that gotcha exist _because_ we're bad at estimating, or do we get bad at estimating because that exists?

**Comment by bretweinraub** [comment]
>  Here's a big question: does that gotcha exist because we're bad at estimating, or do we get bad at estimating because that exists?

Some kind of game theory at work here......

**Comment by JeffCarpenter** [comment]
Definitely. Let's say a non-technical manager Alice hires Bob to add a feature to her application. She has very little idea of _why_ the feature takes the "man-months" that it does. She could attribute it to her codebase's quality, Bob's competence or his working conditions. But there are a few sources of information she does have: she can ask him (take his word for it) or hire another developer to look at his code (get a second opinion).

Or she can see what difference reducing the budget (# of hours) makes on quality while requesting/demanding the same scope and quality of work. If this tactic works and neither scope nor quality slip, she's discovered producer surplus and has now managed to reduce (or eliminate it). If it doesn't, she's sabotaged her own application.
